EveryBeam/wsclean incompatibility in the singularity

Hi, I tried to build a singularity with the recipe in the repository, but it crashes while compiling wsclean:

```
......
-- Found EveryBeam 0.5.1.
--   EveryBeam include dir: /usr/local/include
--   EveryBeam lib: /usr/local/lib/libeverybeam.so
-- EveryBeam version is incompatible
CMake Error at CMakeLists.txt:180 (message):
  WSClean needs EveryBeam version 0.5.x with x>=2 - but found version 0.5.1


-- Configuring incomplete, errors occurred!
See also "/usr/local/src/wsclean/build/CMakeFiles/CMakeOutput.log".
See also "/usr/local/src/wsclean/build/CMakeFiles/CMakeError.log".
FATAL:   While performing build: while running engine: exit status 1
```

I guess that we stick to an old EveryBeam because of the usage of a specific DP3 version, right?

 https://github.com/mhardcastle/ddf-pipeline/blob/c42e8c682e5376a2e33da655b9bc16758ce75432/ddf-py3.singularity#L116

If not possible to move forward with DP3 (and thus with EveryBeam), the recipe should point to an old wsclean to have a working sigularity. The latest wsclean is not required to run facetselfcal on LoTSS data..

https://github.com/mhardcastle/ddf-pipeline/blob/c42e8c682e5376a2e33da655b9bc16758ce75432/ddf-py3.singularity#L147

Update: I tried to install the latest released version of wsclean (v3.3), but in this case EveryBeam v0.5.1 is too recent:

```
...
-- Found EveryBeam 0.5.1.
--   EveryBeam include dir: /usr/local/include
--   EveryBeam lib: /usr/local/lib/libeverybeam.so
-- EveryBeam version is incompatible
CMake Error at CMakeLists.txt:180 (message):
  WSClean needs EveryBeam version 0.4.x with x>=0 or 0.5.0 - but found
  version 0.5.1


-- Configuring incomplete, errors occurred!
See also "/usr/local/src/wsclean/build/CMakeFiles/CMakeOutput.log".
See also "/usr/local/src/wsclean/build/CMakeFiles/CMakeError.log".
```

Wsclean versions <=3.2 instead depend on EveryBeam <=0.4 (https://gitlab.com/aroffringa/wsclean/-/releases)

--- a/ddf-py3.singularity
+++ b/ddf-py3.singularity
@@ -1,25 +1,87 @@
 Bootstrap: docker
 From: debian:bullseye
 
-
+%post -c /bin/bash
+    CUSTOM_ENV=/.singularity.d/env/99-zz_custom_env.sh
+    cat >$CUSTOM_ENV <<EOF
+#!/bin/bash
+PS1="\[\e[1;92m\][Singularity]\[\e[m\] \u@\h:\w\$ "
+EOF
+    chmod 755 $CUSTOM_ENV
+    
 %environment
    export SRC=/usr/local/src
    . $SRC/ddf-pipeline/init.sh
    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib/:/usr/local/idg/lib/
+   export SRC=/usr/local/src
+   export KILLMS_DIR=$SRC
+   export PYTHONPATH=$KILLMS_DIR:$PYTHONPATH
+   export PYTHONPATH=$KILLMS_DIR/DDFacet:$PYTHONPATH
+   export PYTHONPATH=$KILLMS_DIR/killMS:$PYTHONPATH
+   export PYTHONPATH=$KILLMS_DIR/DynSpecMS:$PYTHONPATH
+   export PYTHONPATH=$KILLMS_DIR/drawMS:$PYTHONPATH
+   export PYTHONPATH=$KILLMS_DIR/ms_info:$PYTHONPATH
+   
+   export PATH=$KILLMS_DIR/killMS/killMS:$PATH
+   export PATH=$KILLMS_DIR/DDFacet/DDFacet:$PATH
+   export PATH=$KILLMS_DIR/DDFacet/SkyModel:$PATH
+   export PATH=$KILLMS_DIR/DynSpecMS:$PATH
+   export PATH=$KILLMS_DIR/drawMS:$PATH
+   export PATH=$KILLMS_DIR/ms_info:$PATH
+   export LD_LIBRARY_PATH=$KILLMS_DIR/DDFacet/DDFacet/cbuild:$LD_LIBRARY_PATH
+   #   clear
+   
+   python /usr/bin/logo.py
+
+   
+
+%files
+
 
 
 %post
    export DEBIAN_FRONTEND=noninteractive
-   export J=16
+   export J=32
    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib/:/usr/local/idg/lib/
+   export SRC=/usr/local/src
+   
+   $SRC
 
-   echo "Here we are installing software and other dependencies for the container!"
+
+  ##############################
+  # Dev branches names 
+  export BCH_DDF=master
+  export BCH_KMS=master
+  export BCH_DYN=MTilde_ASync_weights_mergeJones_merge
+  export BCH_DDFPIPE=master
+  ##############################
+
+
+
+   ls $SRC
+  echo "Here we are installing software and other dependencies for the container!"
    apt-get update
-   apt-get install -y emacs htop nano mc \
+
+
+
+
+apt install -y boxes
+#apt install -y spyder
+
+apt install -y openssh-client 
+eval `ssh-agent -s`
+ssh-add
+
+
+apt-get install -y git mlocate
+
+
+apt-get install -y emacs htop nano mc \
     git \
     wget \
     rsync \
     python3-pip \
+    vim \
     libfftw3-dev \
     python3-numpy \
     libfreetype6-dev \
@@ -48,29 +110,241 @@ From: debian:bullseye
     python3-lxml \
     python3-pandas \
     python3-psutil \
-    python3-pyfftw python3-pymysql python3-scipy \
-    python3-requests python3-deap \
+    python3-pyfftw python3-pymysql  \
+    python3-requests  \
     python3-sshtunnel \
-    python3-ruamel.yaml python3-ephem \
+    python3-ruamel.yaml python3-ephem python3-photutils \
+    python3-venv \
+    bzip2 \
     ipython3 \
     libgsl-dev \
     libgtkmm-3.0-dev \
-    libcfitsio-bin libxml2-dev libarmadillo-dev libsigc++-2.0-dev liblua5.3-dev    
+    libcfitsio-bin libxml2-dev libarmadillo-dev libsigc++-2.0-dev liblua5.3-dev libxslt-dev tcl zip libbz2-dev
+    #saods9 
    update-alternatives --install /usr/bin/python python /usr/bin/python3.9 1
-   apt-get install -y casacore-dev casacore-tools python3-casacore cmake curl
-   apt-get install -y python3-astlib python3-ipdb python3-nose python3-metaconfig jq util-linux bsdmainutils
+   apt-get install -y casacore-dev casacore-data casacore-tools python3-casacore cmake curl
+   apt-get install -y python3-astlib python3-ipdb python3-nose python3-metaconfig jq util-linux bsdmainutils evince ristretto imagemagick
+   pip3 install certifi
 
-   pip3 install pybind11
-   pip3 install dask codex_africanus ephem Polygon3 pyregion terminal pyephem ptyprocess timeout-decorator astroquery # pyfits
-   pip3 install --ignore-installed numpy==1.19.5
-   pip3 install reproject tqdm
-   export SRC=/usr/local/src
+  # fitsio
+  cd $SRC
+  git clone https://github.com/esheldon/fitsio.git
+  cd fitsio
+  python setup.py install
 
-   # PyBDSF
-   # pip3 install git+https://github.com/lofar-astron/PyBDSF.git
-   pip3 install bdsf # workaround for https://github.com/lofar-astron/PyBDSF/issues/179
 
-   # LOFAR beam -- for DDF
+# ####################
+apt-get install -y tar wget
+
+apt-get install -y saods9
+
+#cd $SRC
+#wget https://ds9.si.edu/download/source/ds9.8.4.1.tar.gz
+#tar -zxvf ds9.8.4.1.tar.gz
+#cd SAOImageDS9
+#./unix/configure
+#make -j $J
+#cp bin/ds9 /usr/bin/.
+#cd $SRC
+# ####################
+
+
+
+
+# pip3 install opencv-python
+
+
+
+   
+  echo
+  echo
+  echo
+  pip3 show numpy scipy
+  echo
+  echo
+  echo
+    pip3 install Pillow==9.3.0
+  pip3 show numpy scipy
+    pip3 install imageio==2.4.1
+    pip3 install PyWavelets==1.1.1
+    pip3 install tifffile==2019.7.26
+    pip3 install scikit-image==0.19.3
+    # tifffile, scikit-image, pywavelets, imageio affect numpy version
+  pip3 show numpy scipy
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+  pip3 install reproject==0.10.0
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+   pip3 install aplpy==2.1.0
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+  
+  pip3 install pybind11==2.9.2
+   pip3 install dask codex_africanus==0.2.10 "ephem<=9.99" "Polygon3<=3.0.9.1" pyregion terminal "pyephem<=9.99" "ptyprocess<=0.7.0" timeout-decorator astroquery # pyfits
+   pip3 install --ignore-installed numpy==1.19.5 scipy==1.5.4 deap==1.3.1
+   pip3 install reproject
+   pip3 install meqtrees-cattery==1.7.6
+   pip3 install "regions<=0.5" "pywavelets<=1.1.1" "tqdm<=4.64.0" "configparser<=5.2.0"
+  cd $SRC
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+  # PyBDSF
+  pip3 install git+https://github.com/lofar-astron/PyBDSF.git
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+   # few more DDF dependencies
+   pip3 install -U tables==3.7.0 prettytable==2.5.0 pylru==1.2.1 emcee astropy_healpix "sharedarray<=3.2.1" "Cython<=0.29.30"
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+  # tools
+  cd $SRC
+  git clone https://github.com/IanHeywood/ms_info.git
+
+  
+  
+
+  # DDFacet
+  cd $SRC
+  git clone -b v0.7.2 https://github.com/saopicc/DDFacet.git
+  cd DDFacet
+  git checkout $BCH_DDF
+  cd ..
+  #pip install -e ./DDFacet
+  cd DDFacet
+  #rm -rf DDFacet/cbuild
+  #python setup.py clean
+  python setup.py build
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+  
+  # killMS
+  cd $SRC
+  git clone git@github.com:cyriltasse/killMS.git
+  cd killMS
+  git checkout $BCH_KMS
+  rm -rf killMS/cbuild
+  python setup.py build
+  #  python setup.py install
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+  # killMS
+  cd $SRC
+  git clone git@github.com:cyriltasse/drawMS.git
+
+
+
+
+  # Nenupy
+  cd $SRC
+  pip3 install --upgrade https://github.com/AlanLoh/nenupy/tarball/master
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
+  # dynspecMS
+  cd $SRC
+  git clone git@github.com:cyriltasse/DynSpecMS.git
+  cd DynSpecMS
+  git checkout $BCH_DYN
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
+# LSMTool 
+   cd $SRC
+   git clone https://github.com/darafferty/LSMTool.git
+   cd LSMTool
+   python setup.py install
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+  
+   # LofarStMan
+   cd $SRC
+   git clone https://github.com/lofar-astron/LofarStMan.git
+   cd LofarStMan
+   mkdir cbuild; cd cbuild
+   cmake ..
+   make -j $J
+   make install
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+# LOFAR beam -- for DDF
    cd $SRC
    git clone -b v4.1.1 https://github.com/lofar-astron/LOFARBeam.git
    cd LOFARBeam
@@ -80,10 +354,26 @@ From: debian:bullseye
    make -j $J
    make install
 
-   cd /usr/local/lib/python3.9/dist-packages
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+cd /usr/local/lib/python3.9/dist-packages
    ln -s /usr/local/lib/python3.9/site-packages/lofar
 
-   # dysco -- for DP3
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+# dysco -- for DP3
    cd $SRC
    git clone https://github.com/aroffringa/dysco.git
    cd dysco
@@ -95,15 +385,35 @@ From: debian:bullseye
    make install
    cp $SRC/dysco/build/decompress /usr/local/bin/decompress  # https://github.com/aroffringa/dysco/issues/12#issuecomment-773134161
 
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
    # IDG -- for wsclean and DP3
    cd $SRC
-   git clone -b 1.1.0 https://gitlab.com/astron-idg/idg.git
+   git clone -b 0.8 https://gitlab.com/astron-idg/idg.git
    cd idg && mkdir build && cd build
-   cmake -DCMAKE_INSTALL_PREFIX=/usr/local/idg/ ..
+   #   cmake -DCMAKE_INSTALL_PREFIX=/usr/local/idg/ ..
+   cmake ..
    make -j $J
    make install
 
-   # aoflagger -- for DP3
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
+# aoflagger -- for DP3
    cd $SRC
    git clone -b v3.1.0 https://gitlab.com/aroffringa/aoflagger.git
    cd aoflagger
@@ -113,97 +423,223 @@ From: debian:bullseye
    make -j $J
    make install
 
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
    # Everybeam -- for DP3
    cd $SRC
-   git clone -b v0.5.1 https://git.astron.nl/RD/EveryBeam.git
+   #git clone -b v0.1.3 https://git.astron.nl/RD/EveryBeam.git
+   git clone https://git.astron.nl/RD/EveryBeam.git
    cd EveryBeam
+   #   git checkout d75cfa914aa84c556467298237333c2308b592b5
    mkdir build
    cd build
    cmake -DBUILD_WITH_PYTHON=On ..
    make -j $J
    make install
 
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
    # DP3
    cd $SRC
+   #git clone -b v5.1 https://github.com/lofar-astron/DP3.git
    git clone https://github.com/lofar-astron/DP3.git
    cd DP3
    mkdir build
    cd build
-   cmake .. -DCMAKE_PREFIX_PATH=/usr/local/idg
+   #   cmake .. -DCMAKE_PREFIX_PATH=/usr/local/idg
+   cmake ..
    make -j $J
    make install
 
-   # few more DDF dependencies
-   pip3 install -U tables prettytable pylru emcee astropy_healpix sharedarray
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
+
 
    # losoto -- for selfcal
    pip3 install losoto
 
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
    # APLpy -- for selfcal
    pip3 install pyavm
-   pip3 install imageio==2.14.1
-   pip3 install aplpy
 
-  # wsclean latest -- for selfcal
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
+
+pip3 install imageio==2.14.1
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+  
+   pip3 install aplpy==2.1.0
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
+   # wsclean latest -- for selfcal
    cd $SRC
+   #   git clone -b v3.0 https://gitlab.com/aroffringa/wsclean.git
    git clone https://gitlab.com/aroffringa/wsclean.git
    cd wsclean
    mkdir -p build
    cd build
-   cmake .. -DCMAKE_PREFIX_PATH=/usr/local/idg
+   #cmake .. -DCMAKE_PREFIX_PATH=/usr/local/idg
+   cmake ..
    make -j $J
    make install
-
-  # DDFacet
-  cd $SRC
-  git clone -b v0.6.0 https://github.com/saopicc/DDFacet.git
-  cd DDFacet
-  python setup.py install
-
-  # killMS
-  cd $SRC
-  git clone -b v3.0.1 https://github.com/saopicc/killMS.git
-  cd killMS
-  python setup.py install
-
-  # dynspecMS
-  cd $SRC
-  git clone https://github.com/cyriltasse/DynSpecMS.git
 
   # lotss-query
   cd $SRC
   git clone https://github.com/mhardcastle/lotss-query.git
 
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
   # lotss-hba-survey (not needed for most users)
   cd $SRC
   git clone https://github.com/mhardcastle/lotss-hba-survey.git
 
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
   # ddf-pipeline
   cd $SRC
   git clone https://github.com/mhardcastle/ddf-pipeline.git
+  cd ddf-pipeline
+  git checkout $BCH_DDFPIPE
+  cd ..
   # create the init script
   ddf-pipeline/scripts/install.sh
+  cp ddf-pipeline/bin/logoSingularity.py /usr/bin/logo.py
+  
+  # git config --global --add safe.directory /usr/local/src/ddf-pipeline
+  
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
 
   cd /usr/local/src
   wget https://rclone.org/install.sh
   bash install.sh
+
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
 
   cd /usr/local/src
   git clone https://github.com/sara-nl/SpiderScripts.git
   cd SpiderScripts
   cp ada/ada /usr/local/bin
 
+  echo
+  echo
+  echo
+  pip show numpy scipy
+  echo
+  echo
+  echo
+
+
   #pip list --format=columns
   pip3 list --format=columns
+
+
 
   pip3 cache purge
   #apt-get purge -y cmake
   apt-get -y autoremove
   rm -rf /var/lib/apt/lists/*
-  
+  rm -rf /root/.ssh
   bash -c "rm -rf /usr/local/src/{DP3,EveryBeam,LOFARBeam,aoflagger,dysco,idg,wsclean,PyBDSF,SpiderScripts}/" # DDFacet,killMS
-  #ln -s /usr/local/bin/DPPP /usr/local/bin/DP3
+
+
+  echo
+  echo
+  echo
+  pip3 show numpy scipy
+  echo
+  echo
+  echo
+
+
+#  ln -s /usr/local/bin/DPPP /usr/local/bin/DP3
 
 # %runscript
 #   echo source $SRC/ddf-pipeline/init.sh >> $HOME/.bashrc    # not working
 #   . $SRC/ddf-pipeline/init.sh                               # not working
+
+
+# %startscript
+
+# %startscript
+# . /root/set-alias.sh
+# #   alias ip=ipython3
+#    eval ${@}

--- a/ddf-py3.singularity
+++ b/ddf-py3.singularity
@@ -529,7 +529,7 @@ pip3 install imageio==2.14.1
   cd ..
   # create the init script
   ddf-pipeline/scripts/install.sh
-  cp ddf-pipeline/bin/logoSingularity.py /usr/bin/logo.py
+  cp ddf-pipeline/misc/logoSingularity.py /usr/bin/logo.py
   
   # git config --global --add safe.directory /usr/local/src/ddf-pipeline
   

--- a/ddf-py3.singularity
+++ b/ddf-py3.singularity
@@ -243,9 +243,6 @@ apt-get install -y saods9
   # DDFacet
   cd $SRC
   git clone -b v0.7.2 https://github.com/saopicc/DDFacet.git
-  cd DDFacet
-  git checkout $BCH_DDF
-  cd ..
   #pip install -e ./DDFacet
   cd DDFacet
   #rm -rf DDFacet/cbuild
@@ -263,46 +260,10 @@ apt-get install -y saods9
   
   # killMS
   cd $SRC
-  git clone git@github.com:cyriltasse/killMS.git
+  git clone https://github.com/saopicc/killMS.git
   cd killMS
-  git checkout $BCH_KMS
   rm -rf killMS/cbuild
   python setup.py build
-  #  python setup.py install
-
-  echo
-  echo
-  echo
-  pip show numpy scipy
-  echo
-  echo
-  echo
-
-  # killMS
-  cd $SRC
-  git clone git@github.com:cyriltasse/drawMS.git
-
-
-
-
-  # Nenupy
-  cd $SRC
-  pip3 install --upgrade https://github.com/AlanLoh/nenupy/tarball/master
-
-  echo
-  echo
-  echo
-  pip show numpy scipy
-  echo
-  echo
-  echo
-
-
-  # dynspecMS
-  cd $SRC
-  git clone git@github.com:cyriltasse/DynSpecMS.git
-  cd DynSpecMS
-  git checkout $BCH_DYN
 
   echo
   echo

--- a/docs/manual.md
+++ b/docs/manual.md
@@ -246,6 +246,10 @@ parameters). Generally the default values should be sensible.
 Controls the compression of some final images and measurement sets. If
 `compress_ms` is `True` then DP3 will be run to compress the output data.
 
+### [spectra]
+
+Dynamical spectra require the DynSpecMS package.
+
 ## running the code
 
 Once you have set up a config file, make sure you are in the directory

--- a/docs/manual.md
+++ b/docs/manual.md
@@ -268,7 +268,7 @@ you. Good images to look at (use a FITS viewer like ds9) are files of
 the form `image_*_app.restored.fits`. These are the final imaging
 products, in apparent flux units, for each of the self-calibration
 steps. When the code is finished,
-`image_full_ampphase1m.int.restored.fits` is the full-resolution image
+`image_full_ampphase_di_m.NS.app.restored.fits` is the full-resolution image
 in physical units (unless you have specified correction for offsets,
 see below). A file `summary.txt` is also created summarizing the
 run. Log files for the individual steps are stored by default in the

--- a/docs/singularity.md
+++ b/docs/singularity.md
@@ -1,0 +1,26 @@
+# building the singularity image
+
+This is the recommended way of installing ddf-pipeline and all
+dependencies. This includes DP3 and wsclean as they are used in the
+self-calibration scripts. It also pulls in the LoTSS repositories
+`lofar-hba-survey` and `lotss-query` as production use of ddf-pipeline
+requires them.
+
+## you will need
+
+* A recent version of singularity
+* The recipe in the root directory of this repository
+
+## to build
+
+* Copy the recipe to a build directory -- you do not need any of the rest of the repository.
+* Identify a large temporary directory for singularity's working files, say `/my/big/file/system`
+* Run `export SINGULARITY_CACHEDIR=/my/big/file/system ; export SINGULARITY_TMPDIR=/my/big/file/system`
+* Run `singularity build --fakeroot ddf.sif ddf-py3.singularity`
+
+(You either need to be root or to have 'fakeroot' permission set up. If the former, remove the `--fakeroot` option above.)
+
+## to run
+
+``singularity shell -B /my/data/directory ddf.sif`` will give you an environment inside the singularity, binding in the file systems or directories you plan to work in. Note that to run ddf-pipeline itself you will also need the catalogues used for calibration, masking and bootstrap work. Set the environment variable `DDF_PIPELINE_CATALOGS` to the location of your catalogues.
+

--- a/misc/logoSingularity.py
+++ b/misc/logoSingularity.py
@@ -1,0 +1,250 @@
+import os
+try:
+    import DDFacet.Other.ModColor as MC
+    def Str(*args,**kwargs):
+        return MC.Str(*args,**kwargs)
+except:
+    def Str(ss,**kwargs):
+        return ss
+
+
+import shutil
+
+from subprocess import check_output
+from subprocess import Popen, PIPE
+import subprocess
+
+def getVersion(Name):
+    try:
+        r=getVersionSubprocess(Name)
+        if isinstance(r,list):
+            r=" ".join(r)
+            if "ModuleNotFoundError" in r:
+                return "ModuleNotFound"
+        return r.strip()
+                
+    except Exception as e:
+        print('%s: failed with: %s'%(Name, str(e)))
+
+def OS(s):
+    if isinstance(s,list):
+        s=(" ".join(s))
+    os.system("%s > /tmp/OS.log 2>&1"%s)
+    return open("/tmp/OS.log","r").readlines()
+        
+def getVersionSubprocess(Name):
+    if Name=="losoto":
+        command=["losoto", "--version"]
+        #out = check_output(command)
+        out=OS(command)
+        if "not found" in out[0]: return "Not installed"
+        return out[0].strip()
+    elif Name=="wsclean":
+        command=["wsclean","--version"]
+        out=OS(command)
+        #result = subprocess.run(command, capture_output=True, text=True)
+        #stderr,stdout=result.stderr,result.stdout
+        out="".join(out)
+        if "not found" in out[0]: return "Not installed"
+        
+        v=(out.replace("\n"," ").split(" WSClean version ")[1].split("This software package")[0]).replace(" ","")
+        return v
+    elif Name=="DP3" or Name=="aoflagger":
+        command=[Name,"--version"]
+        #result = subprocess.run(command, capture_output=True, text=True)
+        #stderr,stdout=result.stderr,result.stdout
+        out=OS(command)
+        if "not found" in out[0]: return "Not installed"
+        out="".join(out)
+        v=(out.strip().lower().split(Name.lower())[1].replace(" ",""))
+        return v
+    elif Name=="DDF":
+        command=["%s.py"%Name,"--version"]
+        #result = subprocess.run(command, capture_output=True, text=True)
+        #stderr,stdout=result.stderr,result.stdout
+        r=OS(command)
+        if "not found" in r[0]: return "Not installed"
+        # print("rr",r)
+        # #v= stdout.split("\n")[0].split("DDFacet version is ")[1]
+        # print(r[-1].split("DDFacet version is "))
+        rr="Not installed"
+        for l in r:
+            if 'DDFacet version is' in l:
+                rr=l.split("DDFacet version is ")[1].strip()
+                break
+        
+        return rr
+    elif Name=="kMS":
+        command=["%s.py"%Name,"--version"]
+        # result = subprocess.run(command, capture_output=True, text=True)
+        # stderr,stdout=result.stderr,result.stdout
+        # return stdout.split("\n")[-2]
+        r=OS(command)
+        if "not found" in r[0]: return "Not installed"
+        #print("rr",r)
+        #v= stdout.split("\n")[0].split("DDFacet version is ")[1]
+        return r[-1].strip()
+    elif Name=="DynSpecMS":
+        command=["ms2dynspec.py","--version"]
+        result = subprocess.run(command, capture_output=True, text=True)
+        stderr,stdout=result.stderr,result.stdout
+        return stdout.split("\n")[-2].split("version ")[-1]
+    elif Name=="LOFARBeam":
+        command=["python","-c",'"import lofar.stationresponse"']
+        # print(command)
+        # result = subprocess.run(command, capture_output=True, text=True)
+        # stderr,stdout=result.stderr,result.stdout
+
+        s=(" ".join(command))
+        r=OS(s)
+        if len(r)==0:
+            r="Installed, no version available"
+        return r
+    elif Name=="nenupy":
+        command=["python","-c",'"import nenupy,sys; print(str(nenupy.__version__),file=sys.stdout)"']
+        s=(" ".join(command))
+        r=OS(s)
+        return r
+    #o=os.popen(s).read().strip()
+        
+    #    return o
+    elif Name=="lsmtool" or Name=="LofarStMan":
+        command=["python","-c",'"import %s,sys"'%Name]
+        s=(" ".join(command))
+        #o=os.popen(s).read().strip()
+        r=OS(s)
+        if len(r)==0:
+            r="Installed, no version available"
+        return r
+    elif Name=="drawMS":
+        command=["drawMS.py","--version"]
+        r=OS(command)
+        if "not found" in r[0]: return "Not installed"
+        # print("rr",r)
+        # #v= stdout.split("\n")[0].split("DDFacet version is ")[1]
+        # print(r[-1].split("DDFacet version is "))
+        rr="Not installed"
+        for l in r:
+            if 'drawMS.py version' in l:
+                rr=l.split("drawMS.py version ")[1].strip()
+                break
+        return rr
+    elif Name=="lotss-query":
+        command=["python","-c",'"import surveys_db"']
+        s=(" ".join(command))
+        r=OS(s)
+        if len(r)==0:
+            r="Installed, no version available"
+        return r
+    elif Name=="ddf-pipeline":
+        command=["python","-c",'"from pipeline_version import version; print(version())"']
+        s=(" ".join(command))
+        r=OS(s)
+        if len(r)==0:
+            r="Installed, no version available"
+        return r
+    else:
+        print(Name)
+
+        
+        
+DicoExe={"losoto":{"Type":"exe",
+                   "Name":"self"},
+         "DP3":{"Type":"exe",
+                "Name":"self"},
+         "wsclean":{"Type":"exe",
+                    "Name":"self"},
+         "aoflagger":{"Type":"exe",
+                      "Name":"self"},
+         "DDF":{"Type":"exe",
+                "Name":"DDF.py"},
+         "kMS":{"Type":"exe",
+                "Name":"kMS.py"},
+         "DynSpecMS":{"Type":"exe",
+                      "Name":"ms2dynspec.py"},
+         "LOFARBeam":{"Type":"Package",
+                      "Name":"lofar.stationresponse"},
+         "nenupy":{"Type":"Package",
+                   "Name":"nenupy"},
+         "lsmtool":{"Type":"exe",
+                   "Name":"self"},
+         "LofarStMan":{"Type":"Package",
+                       "Name":"LofarStMan"},
+         "drawMS":{"Type":"exe",
+                   "Name":"self"},
+         "lotss-query":{"Type":"Package",
+                        "Name":"surveys_db"},
+         "ddf-pipeline":{"Type":"Package",
+                         "Name":"run_full_field_reprocessing_pipeline"}
+}
+
+SHIFT=30
+
+def Print_v():
+    for Name in DicoExe.keys():
+
+        D=DicoExe[Name]
+        
+        version=getVersion(Name)#.rjust(30," ")
+        NameS=Name#.rjust(20," ")
+        #print(NameS,version)
+        pLine([NameS,version],SHIFT)
+    
+    # if D["Type"]=="exe":
+    #     exeName=D["Name"]
+    #     if exeName=="self":
+    #         exeName=Name
+    #     Path=shutil.which(exeName)
+    #     if Path is None:
+    #         Status=("Not installed")
+    #     else:
+    #         Status=("Installed    ")
+    # else:
+    #     try:
+    #         exec("import %s"%D["Name"])
+    #         Status=("Installed    ")
+    #     except:
+    #         Status=("Not installed")
+    # ss="%20s : %s"%(Name,Status)
+
+
+W=90
+l="%"+"%is"%(W-5)
+ll="  |%s|"%(l)
+def pLine(s,justify="center",length=None):
+    if justify=="center":
+        #print(ll%(str(s).center(W-5," ")))
+        ls=len(s)
+        if length is not None:
+            ls=length
+        
+        S=" "*((W-ls-5)//2)+s+" "*((W-ls-5)//2)
+        S0=" "*((W-ls-5)//2)+" "*ls+" "*((W-ls-5)//2)
+        if len(S0)%2==0: S+=" "
+        print("  |%s|"%S)
+        
+        #print(ll%(str(s).center(W-5," ")))
+    elif isinstance(s,list):
+        s0,s1=s
+        #print(s0,s1)
+        #print(s0.rjust(justify," "))
+        ss=(s0.rjust(justify," ")+" : "+s1)
+        #print(ss,len(ss))
+        F=" "*(W-len(ss)-5)
+        print("  |"+ss+F+"|")
+        
+Sep="="*(W-5)
+print()
+pLine(Sep)
+s="Radio soft singularity image"
+pLine(Str(s.upper(),Bold=1,col="green"),length=len(s))
+#pLine("Radio soft singularity image".upper())
+pLine("")
+pLine(["To source an external dev dir"," source setDev.sh <DirName>"],SHIFT)
+pLine(["for example"," source setDev.sh /data/$USER/DEV"],SHIFT)
+pLine(Sep)
+#pLine(["<Software>","<Versions>"],SHIFT)
+#Print_v()
+#pLine(Sep)
+print()
+print()

